### PR TITLE
OpenSSL 1.1 compatibility

### PR DESCRIPTION
**Pull request by welwood08**
Hi, please find here the changes necessary to compile Csocket against OpenSSL 1.1 without errors or deprecation warnings. I've also successfully compiled against 1.0.2h and 0.9.8g, all on Debian, as part of my solution to znc/znc#1310. I have not yet exercised any of the changed code in a running application.

The commits are currently fairly granular in order to solicit feedback, however if desired I can squash and force push after addressing any comments. I'm not very experienced with C++ so please do let me know if I've made a glaring mistake or there's a better convention for something I've done, etc.

### CODE REVIEW

File: Csocket.cc [hunk 1]
```diff
@@ -2691,39 +2746,56 @@ CS_STRING Csock::GetPeerPubKey() const
 {
 	CS_STRING sKey;
 
-	SSL_SESSION * pSession = GetSSLSession();
+	X509 * pCert = GetX509();
```
Comment (DarthGandalf) [inline Csocket.cc]: I think it makes this method to return local key instead of remote key...

File: Csocket.cc [hunk 2]
```diff
@@ -2691,39 +2746,56 @@ CS_STRING Csock::GetPeerPubKey() const
...
+#else
+			int iType = pKey->type;
+#endif /* HAVE_OPAQUE_EVP_PKEY */
+			switch( iType )
 			{
+#ifndef OPENSSL_NO_RSA
 			case EVP_PKEY_RSA:
-			{
-				hxKey = BN_bn2hex( pKey->pkey.rsa->n );
+# ifdef HAVE_OPAQUE_RSA
```
Comment (DarthGandalf) [inline Csocket.cc]:
```c++
#ifndef ...
# define EVP_PKEY_get0_RSA(pKey) ((pKey)->pkey.rsa)
#endif
```

Something similar for `RSA_get0_key` but as inline function. `EVP_PKEY_get0_RSA` also can be defined as inline function instead.

Was there any OpenSSL version released which has one opaque struct, but not the other? If not, I don't think it's necessary to support such case, which would simplify code.

**Comment by jimloco** [comment]
Given the nature of this change and what it may affect, we would need a new regression test (in the examples sub directory) to accompany this commit in order to verify that existing functionality remains unchanged.

Thanks.

File: Csocket.cc (hunk 1 above)
Comment (welwood08) [inline Csocket.cc]: The old code with GetSSLSession calls [SSL_get_session](https://github.com/openssl/openssl/blob/a230b26e0959dc5f072fbbdadcc9ed45e904c50c/ssl/ssl_sess.c#L47-L51) which returns `ssl->session`. Later the code accesses the `peer` property of that object.

GetX509 calls [SSL_get_peer_certificate](https://github.com/openssl/openssl/blob/cfd20f64cc4bd440cfc8fe59f2daaa575015af3d/ssl/ssl_lib.c#L1322-L1337) which you can see directly returns `ssl->session->peer` (bypassing the now opaque session object).

The only meaningful difference here is that now an internal reference count is incremented on the cert, hence it needs to be freed after use.

**Comment by welwood08** [comment]
Having looked in the examples directory for the first time yesterday, I'll admit I hadn't imagined it would contain tests. While I'm reasonably comfortable hacking on existing code, writing something from scratch to test what I've hacked is a pretty daunting prospect at my current skill level in the language. Not to mention, the existing tests don't compile on either of my available systems out of the box (some simple Makefile tweaks seem to work on one though). This may take some time.

**Comment by welwood08** [comment]
Hi, just checking in because "a while" has passed - I haven't given up!

I've been chipping away at writing something that tests all the bits of Csocket touched by my changes, and simultaneously learning what I need to accomplish that (plus a lot that it turned out I didn't need).

Current progress:
- [x] Tweak makefile to get builds working (most significantly: conditionally link objects based on source dependencies, pulling in unused system curl was preventing me from testing multiple openssl versions)
- [x] Csocket as client: verify SHA1 fingerprint returned by Csocket matches expectation
- [x] Csocket as client: verify public key (RSA modulus) returned by Csocket matches expectation
- [x] Csocket as client: verify SSL protocol version of connection matches expectation
- [x] Csocket as server: verify SSL protocol version of connection matches expectation
- [ ] Csocket as server: verify client public key and fingerprint?
- [ ] DSA?
- [ ] ensure server/client protocol version mismatches result in failed connection?
- [ ] try to fix failures and unexpected results of SSLv3 and TLSv1x tests with Csocket as server built with openssl 0.9.8? (at least one "unexpected" result is due to the bug in existing code that I identified in 448e18a)

I'm worried that I've already breached 500 lines of code and while some bloat may be due to me bodging things together as I learn, it still feels like too much (in terms of what people might want to review)... Is what I already have going to be enough of a regression test? On my system, all my tests pass both before and after my changes to Csocket when built with openssl 1.0.2, and after my changes when built with openssl 1.1.0.

**Comment by lll2086** [comment]
Given that it is a major update,  it is good to have a comprehensive regression test of the changes in order to verify functionality. Don't worry too much about the code size or time.

Thanks

**Comment by lll2086** [comment]
We are going to pull your changes and refractor if needed

**Comment by welwood08** [comment]
Ah cool, thanks. I was faffing around trying to get OpenSSL 1.0.0 to compile and pass tests when I last spent time on this. Would my regression test still be welcome? If so, presumably another pull req?

**Comment by lll2086** [comment]
No problem. The regression test is still welcome. It's always nice to have more tests. Yeah, it'll probably be another pull request.

**Pull request merged**